packages/core/src/external/athenahealth/index.ts

```diff
@@ -397,6 +437,238 @@ class AthenaHealthApi {
+        this.s3Utils
+          .uploadFile({
+            bucket: responsesBucket,
+            key,
+            file: Buffer.from(JSON.stringify(response.data), "utf8"),
+            contentType: "application/json",
+          })
+          .catch(processAsyncError("Error saving to s3 @ AthenaHealth - createProblem"));
+      }
+      const outcome = problemCreateResponseSchema.parse(response.data);
+      if (!outcome.success) throw new MetriportError(`Problem create not successful`);
+      return outcome;
+    } catch (error) {
+      const msg = `Failure while creating problem @ AthenaHealth`;
+      log(`${msg}. Cause: ${errorToString(error)}`);
+      capture.error(msg, {
+        extra: {
+          url: chartProblemUrl,
+          cxId,
+          practiceId: this.practiceId,
+          patientId,
+          departmentId,
+          context: "athenahealth.create-problem",
+          error,
+        },
+      });
+      throw error;
+    }
+  }
+
+  async createVitals({
+    cxId,
+    patientId,
+    departmentId,
+    vitals,
+  }: {
+    cxId: string;
+    patientId: string;
+    departmentId: string;
+    vitals: GroupedVitals;
+  }): Promise<VitalsCreateResponse[]> {
+    const { log, debug } = out(
+      `AthenaHealth create vitals - cxId ${cxId} practiceId ${this.practiceId} patientId ${patientId} departmentId ${departmentId}`
+    );
+    const chartVitalsUrl = `/chart/${this.stripPatientId(patientId)}/vitals`;
+    try {
+      const observation = vitals.mostRecentObservation;
+      const code = this.getObservationCode(observation);
+      if (!code) {
+        throw new MetriportError("No code found for observation", undefined, {
+          cxId,
+          practiceId: this.practiceId,
+          patientId,
+          departmentId,
+          observationId: observation.id,
+        });
+      }
+      const clinicalElementId = vitalSignCodesMapAthena.get(code);
+      if (!clinicalElementId) {
+        throw new MetriportError("No clinical element id found for observation", undefined, {
+          cxId,
+          practiceId: this.practiceId,
+          patientId,
+          departmentId,
+          observationId: observation.id,
+        });
+      }
+      if (!vitals.sortedPoints || vitals.sortedPoints.length === 0) {
+        throw new MetriportError("No points found for vitals", undefined, {
+          cxId,
+          practiceId: this.practiceId,
+          patientId,
+          departmentId,
+          observationId: observation.id,
+        });
+      }
+      const payloads = vitals.sortedPoints.map(v => {
+        const vitalsData = this.createVitalsData(v, clinicalElementId);
+        if (uniqBy(vitalsData, "readingtaken").length !== vitalsData.length) {
+          throw new MetriportError("Duplicate reading taken for vitals", undefined, {
+            cxId,
+            practiceId: this.practiceId,
+            patientId,
+            departmentId,
+            observationId: observation.id,
```

Comment:
This is written multiple times, maybe create some sort of "context" object?

Reply:
Done.